internal/gps/pkgtree/digest.go

```diff
@@ -0,0 +1,442 @@
+// solidus character, `/`, as their path separator. For example,
+// "github.com/alice/alice1".
+func VerifyDepTree(vendorRoot string, wantSums map[string][]byte) (map[string]VendorStatus, error) {
+	vendorRoot = filepath.Clean(vendorRoot) + pathSeparator
+
+	// NOTE: Ensure top level pathname is a directory
+	fi, err := os.Stat(vendorRoot)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot Stat")
+	}
+	if !fi.IsDir() {
+		return nil, errors.Errorf("cannot verify non directory: %q", vendorRoot)
+	}
+
+	var otherNode *fsnode
+	currentNode := &fsnode{pathname: vendorRoot, parentIndex: -1, isRequiredAncestor: true}
+	queue := []*fsnode{currentNode} // queue of directories that must be inspected
+	prefixLength := len(vendorRoot)
+
+	// In order to identify all file system nodes that are not in the lock file,
+	// represented by the specified expected sums parameter, and in order to
+	// only report the top level of a subdirectory of file system nodes, rather
+	// than every node internal to them, we will create a tree of nodes stored
+	// in a slice.  We do this because we do not know at what level a project
+	// exists at. Some projects are fewer than and some projects more than the
+	// typical three layer subdirectory under the vendor root directory.
+	//
+	// For a following few examples, assume the below vendor root directory:
+	//
+	// github.com/alice/alice1/a1.go
+	// github.com/alice/alice2/a2.go
+	// github.com/bob/bob1/b1.go
+	// github.com/bob/bob2/b2.go
+	// launghpad.net/nifty/n1.go
+	//
+	// 1) If only the `alice1` and `alice2` projects were in the lock file, we'd
+	// prefer the output to state that `github.com/bob` is `NotInLock`.
+	//
+	// 2) If `alice1`, `alice2`, and `bob1` were in the lock file, we'd want to
+	// report `github.com/bob/bob2` as `NotInLock`.
+	//
+	// 3) If none of `alice1`, `alice2`, `bob1`, or `bob2` were in the lock
+	// file, the entire `github.com` directory would be reported as `NotInLock`.
+	//
+	// Each node in our tree has the slice index of its parent node, so once we
+	// can categorically state a particular directory is required because it is
+	// in the lock file, we can mark all of its ancestors as also being
+	// required. Then, when we finish walking the directory hierarchy, any nodes
+	// which are not required but have a required parent will be marked as
+	// `NotInLock`.
+	nodes := []*fsnode{currentNode}
+
+	// Mark directories of expected projects as required. When the respective
+	// project is found in the vendor root hierarchy, its status will be updated
+	// to reflect whether its digest is empty, or whether or not it matches the
+	// expected digest.
+	status := make(map[string]VendorStatus)
+	for pathname := range wantSums {
+		status[pathname] = NotInTree
+	}
+
+	for len(queue) > 0 {
+		// pop node from the queue (depth first traversal, reverse lexicographical order inside a directory)
+		lq1 := len(queue) - 1
+		currentNode, queue = queue[lq1], queue[:lq1]
```

Comment:
This reminds me of GNU coding standards I read a while back that said not to bother releasing heap allocations just prior to exiting a C program. However, I like my C code to have clean `valgrind` runs, so I always do so anyway. :)

Possible alternate final loop:

```Go
for len(nodes) > 1 {
	// pop off right-most node from slice of nodes
	ln1 := len(nodes) - 1
	currentNode, nodes = nodes[ln1], nodes[:ln1]

	if !currentNode.isRequiredAncestor && nodes[currentNode.parentIndex].isRequiredAncestor {
		status[filepath.ToSlash(currentNode.pathname[prefixLength:])] = NotInLock
	}
}
currentNode, nodes = nil, nil
```

This method pops each node off the slice one by one to walk from the last node to the first, and it overwrites `currentNode` each time, releasing what `currentNode` had been pointing to. On the first iteration, it no longer points to the final node visited in the top loop. For subsequent iterations it releases the previous node popped from the slice when it pops a new value into `currentNode`.

I'll go ahead and make this change, because I think the code is more easy to understand. Please let me know if I have not properly addressed your concerns.